internal/dcache/buffer.go

```diff
@@ -43,26 +44,68 @@ import (
 
 //go:generate $ASSERT_REMOVER $GOFILE
 
-type bufferPool struct {
+var BufPool *BufferPool
+
+type BufferPool struct {
 	pool       sync.Pool    // sync.Pool to relieve GC
 	bufSize    int          // size of buffers in this pool
 	maxBuffers int64        // max allocated buffers allowed
 	curBuffers atomic.Int64 // buffers currently allocated
 }
 
-func NewBufferPool(bufSize int, maxBuffers int) *bufferPool {
-	return &bufferPool{
+func CreateBufferPool(bufSize uint64) error {
+	common.Assert(BufPool == nil)
+	//
+	// Size of buffers managed by bufferPool.
+	// This should be equal to the chunk size we support, since each buffer can hold upto one chunk
+	// worth of data.
+	//
+
+	//
+	// Maximum numbers of 'bufSize' buffers can be allocated from the bufferPool.
+	// We should allow sufficiently many buffers to support at least few files being read/written
+	// simultaneously.
+	// Note that only writeChunk uses buffers from this pool while readChunk uses buffers allocated by
+	// thrift and those are not accounted in this.
+	//
+	// TODO: Find out how/if thrift controls those buffers, or does it result in OOM killing of the
+	//       process.
+	//
+	maxBuffers := uint64(1024)
+
+	//
+	// How much percent of the system RAM (available memory to be precise) are we allowed to use?
+	//
+	// TODO: This can be config value.
+	//
+	usablePercentSystemRAM := 50
+
+	//
+	// Allow higher number of maxBuffers if system can afford.
+	//
+	ramMB, err := common.GetAvailableMemoryInMB()
+	if err != nil {
+		return fmt.Errorf("NewFileIOManager: %v", err)
+	}
+
+	// usableMemory in bytes capped by usablePercentSystemRAM.
+	usableMemory := (ramMB * 1024 * 1024 * uint64(usablePercentSystemRAM)) / 100
+	maxBuffers = max(maxBuffers, usableMemory/bufSize)
+
+	BufPool = &BufferPool{
 		pool: sync.Pool{
 			New: func() any {
 				return make([]byte, bufSize)
 			},
 		},
-		bufSize:    bufSize,
+		bufSize:    int(bufSize),
 		maxBuffers: int64(maxBuffers),
 	}
+
+	return nil
 }
 
-func (bp *bufferPool) getBuffer() ([]byte, error) {
+func (bp *BufferPool) GetBuffer() ([]byte, error) {
 	if bp.curBuffers.Load() > bp.maxBuffers {
```

Comment:
let's add a TODO to make this wait for buffers to get freed and only fail after a timeout.